evo/tools/tf_cache.py

```diff
@@ -113,24 +164,39 @@ def from_bag(self, reader: Rosbag1Reader, topic: str = "/tf",
                     raise TfCacheException(
                         f"Expected {SUPPORTED_TF_MSG} message type for topic "
                         f"{tf_topic}, got: {connection.msgtype}")
-                msg = typestore.deserialize_ros1(rawdata, connection.msgtype)
+                if isinstance(reader, Rosbag1Reader):
+                    msg = typestore.deserialize_ros1(rawdata, connection.msgtype)
+                else:
+                    msg = typestore.deserialize_cdr(rawdata, connection.msgtype)
                 for tf in msg.transforms:  # type: ignore
                     # Convert from rosbags.typesys.types to native ROS.
                     # Related: https://gitlab.com/ternaris/rosbags/-/issues/13
-                    stamp = rospy.Time()
-                    stamp.secs = tf.header.stamp.sec
-                    stamp.nsecs = tf.header.stamp.nanosec
-                    tf = TransformStamped(Header(0, stamp, tf.header.frame_id),
-                                          tf.child_frame_id, tf.transform)
+                    native_msg = TransformStamped()
+                    if isinstance(reader, Rosbag1Reader):
+                        native_msg.header.stamp.secs = tf.header.stamp.sec
+                        native_msg.header.stamp.nsecs = tf.header.stamp.nanosec
+                    else:
+                        native_msg.header.stamp.sec = tf.header.stamp.sec
+                        native_msg.header.stamp.nanosec = tf.header.stamp.nanosec
```

Comment:
The `if isinstance(reader, Rosbag1Reader)` should be replaced by `if os.getenv("ROS_VERSION") == "1"` here and in line 249.

The type of the native message is not depending on the input bag type and should be chosen based on the ROS distro version instead. Nice effect of this: it enables ROS 2 users to load data from old ROS 1 bags that they might have sitting around (and vice versa), without having to use a converter like `rosbags-convert` etc. The normal non-TF topics can be already loaded in that way.

Reply:
Ended up checking object attributes in https://github.com/MichaelGrupp/evo/pull/672/commits/633d507d904dd1e9e6cf1e1165e12488b927f15e rather than hoping an envvar is properly set. PTAL.